rust/worker/src/index/fulltext/types.rs

```diff
@@ -259,18 +303,240 @@ mod test {
+        assert!(res.contains(&4));
 
-        let res = index.search(".!").unwrap();
+        let res = index.search(".!.").unwrap();
         assert!(res.contains(&3));
         assert!(res.contains(&4));
     }
+
+    use proptest::prelude::*;
+    use proptest_state_machine::{prop_state_machine, ReferenceStateMachine, StateMachineTest};
+    use proptest::test_runner::Config;
+    use rand::prelude::IteratorRandom;
+    // https://tonsky.me/blog/unicode/
+    use stringslice::StringSlice;
+
+    #[derive(Debug, Clone)]
+    pub(crate) enum Transition {
+        BeginTransaction,
+        CommitTransaction,
+        AddDocument(String, i32),
+        Search(String),
+    }
+
+    #[derive(Debug, Clone)]
+    pub(crate) struct ReferenceState {
+        in_transaction: bool,
+        // TODO use a real, tested, external index. This works for now but test
+        // time grows quadratically in # of steps, limiting how thorough we
+        // can be.
+        state: HashMap<String, i32>,
+    }
+
+    impl ReferenceState {
+        fn new() -> Self {
+            ReferenceState {
+                in_transaction: false,
+                state: HashMap::new(),
+            }
+        }
+
+        fn search(&self, query: &str) -> Vec<i32> {
+            if query.len() < 3 {
+                return vec![];
+            }
+            if self.in_transaction {
+                return vec![];
+            }
+            let mut results = vec![];
+            for (doc, id) in &self.state {
+                if doc.contains(query) {
+                    results.push(*id);
+                }
+            }
+            results
+        }
+
+        fn add_document(&mut self, doc: String, id: i32) {
+            if doc.len() < 5 {
+                return;
+            }
+            if id < 0 {
+                return;
+            }
+            if !self.in_transaction {
+                return;
+            }
+            self.state.insert(doc, id);
+        }
+    }
+
+    impl ReferenceStateMachine for ReferenceState {
+        type State = ReferenceState;
+        type Transition = Transition;
+
+        fn init_state() -> BoxedStrategy<Self::State> {
+            Just(ReferenceState::new()).boxed()
+        }
+
+        fn transitions(state: &ReferenceState) -> BoxedStrategy<Transition> {
+            // Grab a random chunk of a random string in state
+            let doc = state.state.keys().choose(&mut rand::thread_rng());
+            if doc.is_none() {
+                return prop_oneof![
+                    Just(Transition::BeginTransaction),
+                    (".{4,16000}", (-5..999999)).prop_map(move |(doc, id)| Transition::AddDocument(doc, id)),
+                    ".*{3,1000}".prop_map(Transition::Search),
+                ].boxed();
+            }
+
+            let doc = doc.unwrap();
+            let start = rand::thread_rng().gen_range(0..(doc.len() / 2));
+            let end = rand::thread_rng().gen_range((start + 5)..doc.len());
```

Comment:
does proptest seed tests? should we use something besides thread_rng -- especially for reproducibilty?